Add New Project Tutorial (#190)
Oneflow-Inc/libai: docs/source/tutorials/Build_New_Project_on_LiBai.md

# Build New Project on LiBai

Here we provide the basic guide for users to build new projects based on LiBai, and introduce how to use LiBai to achieve the least code work.

The advantages of using LiBai to start a new project(such as paper reproduction and finetune task) are as follows:

> - Avoid redundant work, developers can directly inherit many bulit-in modules from LiBai.

Reviewer: ```suggestion
- Avoid redundant work, developers can directly inherit many built-in modules from LiBai.
```

Author: 已修改

- `graph`: Import directly, and the model will be automatically converted to graph during operation.

> All imported modules must take LiBai as the root directory, otherwise, the saved `yaml` file will not be able to save the correct path of the module, resulting in an error when reading `yaml`, so the experiment cannot be reproduced.

A complete `config.py` example:

```python
from omegaconf import OmegaConf
from scipy.stats import spearmanr
from configs.common.data.bert_dataset import tokenization
from configs.common.models.bert import cfg as my_cfg
from configs.common.models.graph import graph
from configs.common.optim import optim
from configs.common.train import train
from libai.config import LazyCall
from libai.evaluation import DatasetEvaluator
from libai.data.build import build_nlp_test_loader, build_nlp_train_loader
from libai.tokenizer import BertTokenizer
from libai.optim import get_default_optimizer_params, PolynomialLR
from projects.MyProjects.dataset.dataset import TrainDataset, TestDataset
from projects.MyProjects.modeling import MyModel

def spearman_target(pred, labels):
    # Calculate spearman
    return spearmanr(pred, labels).correlation

> class MyEvaluator(DatasetEvaluator):

Reviewer: 为啥在 config 文件里面定义 Evaluator 呢

Reviewer: 我觉得应该另外有一个 python 文件定义 evaluator，config 去 import 进来使用

Author: 好的 可以的

    dict(
        vocab_size=21128,
        ...
    )
)

model = LazyCall(MyModel)(cfg=my_cfg)

optim = LazyCall(flow.optim.AdamW)(
    parameters=LazyCall(get_default_optimizer_params)(
        # parameters.model is meant to be set to the model object, before instantiating the optimizer.
        clip_grad_max_norm=1.0,
        ...
    ),
    lr=1e-4,
    ...
)

lr_scheduler = LazyCall(flow.optim.lr_scheduler.WarmUpLR)(
    lrsch_or_optimizer=LazyCall(PolynomialLR)(steps=1000, end_learning_rate=1.0e-5,),
    warmup_factor=0,
    ...
)

train.update(
    dict(
        output_dir=".../result",
        train_micro_batch_size=64,
        evaluation=dict(
        enabled=True,
        evaluator=LazyCall(MyEvaluator)(),
            eval_period=5000,
            eval_iter=1e9,
            eval_metric="spearman",
            eval_mode="max",
        ),

        ...
        
        dist=dict(
            data_parallel_size=1,
            tensor_parallel_size=1,
            pipeline_parallel_size=1,
        ),
    )
)

```

After building the `config.py`, if we want to get the corresponding fields in the project, we just need to access like `cfg.my_cfg.***`

## Start Training

> After the above modules are built, we can start training.

Author: 好的 确实不用强调是单卡

```bash
bash projects/my_projects/train.sh projects/my_projects/config.py 1
```

> Config can support both `py` files and generated `yaml` files.

> 

Reviewer 2: 这里顺序有点问题，应该是先展示 train.sh 的文件内容，才是教用户怎么运行脚本，而且你这个脚本里的参数是写死的，没法从命令行获取， https://github.com/Oneflow-Inc/libai/blob/main/tools/train.sh

Author: 好的

Reviewer 3: 可以加一个link https://libai.readthedocs.io/en/latest/tutorials/Train_and_Eval_Command_Line.html
这里有比较详细的脚本启动方式的说明

    - Related class defined with LazyCall which returns a dict instead of calling the object.

2. Prepare a model file(such as [model.py](https://github.com/Oneflow-Inc/libai/blob/main/projects/QQP/modeling/model.py)) which contains:
    - Build related models in this file, the construction method is similar to OneFlow.
    - Because Libai will set up a static diagram by default, the calculation of loss needs to be inside the model.
    - The function `forward` must return a dict.
    - When defining a tensor in the model, you need to use `to_global`, turn tensor into a global pattern.
    - When defining layers, you can import them directly from `libai.layers`, because it have already pre-defined the SBP signature.

3. Prepare a dataset file(such as [dataset.py](https://github.com/Oneflow-Inc/libai/tree/main/projects/QQP/dataset)) which contains:
    - Build `Dataset` in this file, the construction method is similar to OneFlow.
    - The difference is that we need to use `DistTensorData` and `Instance`.
    - The shape of each batch must be global.
    - The return `key`  

>     - In `__getitem__` function, the `key` returned by the method must be global with the parameter name of the 'forward' function in the 'model'.

Reviewer: ```suggestion
    - In `__getitem__` function, the `key` returned by the method must be consistent with the parameter name of the `forward` function in the `model`.
```

Reviewer 4: 这儿确实得consistent，之前consistent改global是描述里面的tensor状态，oneflow已经改成了to_global接口了 😆 ，我的错，不是所有consistent都要改

Author: 好的

## Main Function Entry
[tools/train_net.py](https://github.com/Oneflow-Inc/libai/blob/main/tools/train_net.py) is the default main function entry provided in LiBai.

## Build Config
The `config.py` in LiBai is special, which takes the form of lazyconfig and will be saved as `.yaml` at runtime.
The following describes the complete `config.py` and how to inherit the config in LiBai.

> First, config has several necessary fields:

Reviewer: 因为我们有一章 tutorial 专门讲了 config，可以给他一个链接，下面的描述可以简单一点。

Reviewer 4: 可以 please refer to [config system]()

- `train`: It contains training related parameters and is a dict type.
- `model`: Model used by the task, specify the generation method in the file, due to the characteristics of lazycall, the model     will be generated at runtime.
- `optim`: Optimizer related. Default to AdamW.

> - `lr_scheduler`: Related to learning rate, Default to warmup cosine decay lr-scheduler.

Reviewer 4: `, Default` -> `Related to learning rate. Default to` 句号结尾

> - `graph`: Import directly, and the model will be automatically converted to graph during operation.

Reviewer 4: convert to nn.Graph mode for training and evaluation, please refer to [nn.Graph docs](https://oneflow.readthedocs.io/en/master/graph.html) for more details

from configs.common.models.bert import cfg as my_cfg
from configs.common.models.graph import graph
from configs.common.optim import optim
from configs.common.train import train
from libai.config import LazyCall
from libai.utils import distributed as dist
from libai.data.build import build_nlp_test_loader, build_nlp_train_loader
from libai.tokenizer import BertTokenizer
from libai.optim import get_default_optimizer_params, PolynomialLR
from projects.MyProjects.dataset.dataset import TrainDataset, TestDataset
from projects.MyProjects.modeling import MyModel

tokenization.tokenizer = LazyCall(BertTokenizer)(
    vocab_file=".../vocab.txt",
)

dataloader = OmegaConf.create()
dataloader.train = LazyCall(build_nlp_train_loader)(
    dataset=[
        LazyCall(TrainDataset)(
            path=".../train.txt",
            ...
            ),
        )
    ],
)

dataloader.test = [
    LazyCall(build_nlp_test_loader)(
        dataset=LazyCall(TestDataset)(
            path=".../test.txt",
            ...
            ),
        ),
    ),
]

my_cfg.update(
    dict(
        vocab_size=21128,
        ...
    )
)

model = LazyCall(MyModel)(cfg=my_cfg)

optim = LazyCall(flow.optim.AdamW)(

>     parameters=LazyCall(get_default_optimizer_params)(

Reviewer: ```suggestion
    params=LazyCall(get_default_optimizer_params)(
```

        # parameters.model is meant to be set to the model object, before instantiating the optimizer.
        clip_grad_max_norm=1.0,
        ...
    ),
    lr=1e-4,
    ...
)

> lr_scheduler = LazyCall(flow.optim.lr_scheduler.WarmUpLR)(

Reviewer: lr_scheduler 现在定义在 train.py 里面了

    lrsch_or_optimizer=LazyCall(PolynomialLR)(steps=1000, end_learning_rate=1.0e-5,),
    warmup_factor=0,
    ...
)

train.update(
    dict(
        output_dir=".../result",
        train_micro_batch_size=64,
        evaluation=dict(
        enabled=True,

>         evaluator=LazyCall(MyEvaluator)(),

Reviewer: 这里需要改一下，应该是 evaluation = dict(evaluator=...

            eval_period=5000,
            eval_iter=1e9,
            eval_metric="spearman",
            eval_mode="max",
        ),

        ...
        
        dist=dict(
            data_parallel_size=1,
            tensor_parallel_size=1,
            pipeline_parallel_size=1,
        ),
    )
)

```

After building the `config.py`, if we want to get the corresponding fields in the project, we just need to access like `cfg.my_cfg.***`

## Start Training
The `train.sh` file contains some parameters, such as `GPUS`, `NODE`, etc.

```bash
#!/usr/bin/env bash
FILE=$1
CONFIG=$2
GPUS=$3
NODE=${NODE:-1}
NODE_RANK=${NODE_RANK:-0}
ADDR=${ADDR:-127.0.0.1}
PORT=${PORT:-12345}

python3 -m oneflow.distributed.launch \
--nproc_per_node $GPUS --nnodes $NODE --node_rank $NODE_RANK --master_addr $ADDR --master_port $PORT \
$FILE --config-file $CONFIG ${@:4}
```

After the above modules are built, we can start training with single gpu.

> Config can support both `py` files and generated `yaml` files.

```bash

> bash projects/my_projects/train.sh projects/my_projects/config.py 1

Reviewer: 这里也要改一下，是 `bash projects/my_projects/train.sh train_net.py projects/my_projects/config.py 1`
需要加上训练脚本的路径

# Build New Project on LiBai

Here we provide the basic guide for users to build new projects based on LiBai, and introduce how to use LiBai to achieve the least code work.

The advantages of using LiBai to start a new project(such as paper reproduction and finetune task) are as follows:

> 

Reviewer 4: 前文可以整合成一段话，把`introduce how to use LiBai to achieve the least code work`删了，这一句太奇怪了，你整合成

`Here we provide the basic guide for users to build new projects based on LiBai. The advantages of using LiBai to start a new project(such as paper reproduction and finetune task) are as follows:`

- Avoid redundant work, developers can directly inherit many built-in modules from LiBai.
- Easily reproduce the experiments already run, because LiBai will save the configuration file automatically.
- Automatically output the useful information during training time, such as remaining training time, current iter, throughput, loss information and current learning rate, etc.
- Set a few config params to enjoy distributed training techniques.

## Introduce

> Let's take the [Bert_Finetune](https://github.com/Oneflow-Inc/libai/tree/main/projects/QQP) task as an example to introduce LiBai.

Reviewer 4: `Bert_Finetune`不需要下划线, `BERT finetune`即可

> The complete file structure of a project:

Reviewer 4: `of a project` -> `of the project`

```
projects/my_project
├── configs
│   └── config_custom.py
│   └── ...
├── dataset
│   ├── custom_dataset.py
│   └── ...
├── modeling
│   ├── custom_model.py
│   └── ...
├── README.md
```

> Main steps for starting a new LiBai project:

Reviewer 4: 这里可以改成
`Starting a new project based on LiBai step by step:`, 或者不改也行，看你觉得哪个好

1. Prepare a config file(such as [config.py](https://github.com/Oneflow-Inc/libai/blob/main/projects/QQP/configs/config_qqp.py)) which contains:
    - This file should be independent of the default config in LiBai.

>     - The relevant parameters of the task.

Reviewer 4: 这里有个明显很奇怪的部分: which contains后接的第一点是` this file should be`，太奇怪了，第一点可以直接删了，然后开头改为

Prepare an independent config file (such as [config_qqp.py](https://github.com/Oneflow-Inc/libai/blob/main/projects/QQP/configs/config_qqp.py)) which contains:
- The relevant parameters of the task
- The pre-defined related `Class`

主要的意思就是contains后最好接一个名词或者啥的

    - Defining related Class, such as `Model`, `Optimizer`, `Scheduler`, `Dataset`.
    - You can inherit the default config in `configs/common` and rewrite it, which can greatly reduce the workload.
    - Related class defined with LazyCall which returns a dict instead of calling the object.

> 2. Prepare a model file(such as [model.py](https://github.com/Oneflow-Inc/libai/blob/main/projects/QQP/modeling/model.py)) which contains:

Reviewer 4: such as [qqp model]()

    - Build related models in this file, the construction method is similar to OneFlow.
    - Because Libai will set up a static diagram by default, the calculation of loss needs to be inside the model.
    - The function `forward` must return a dict.

>     - When defining a tensor in the model, you need to use `to_global`, turn tensor into a consistent pattern.

Reviewer 4: turn tensor into global pattern

>     - When defining layers, you can import them directly from `libai.layers`, because it already have SBP defined.

Reviewer 4: 我感觉 it already have SBP defined 很奇怪, 最好改为` it have already pre-defined the SBP signature`类似的

3. Prepare a dataset file(such as [dataset.py](https://github.com/Oneflow-Inc/libai/tree/main/projects/QQP/dataset)) which contains:
    - Build `Dataset` in this file, the construction method is similar to OneFlow.

>     - The difference is that we need to use `DistTensorData` and `Instance`.

Reviewer 4: 这边用到的一些技术比如`DistTensorData`可以强调一下`which will be introduced in Advanced Tutorials`

>     - The shape of each batch must be consistent.

Reviewer 4: consistent都改为global吧，oneflow重命名了这个接口

    - The return `key`  
    - In `__getitem__` function, the `key` returned by the method must be consistent with the parameter name of the 'forward' function in the 'model'.

## Main Function Entry

> [tools.train_net.py](https://github.com/Oneflow-Inc/libai/blob/main/tools/train_net.py) is the default main function entry provided in LiBai.

Reviewer 4: `tools/train_net.py`比较好

## Build Config
The `config.py` in LiBai is special, which takes the form of lazyconfig and will be saved as `.yaml` at runtime.
The following describes the complete `config.py` and how to inherit the config in LiBai.

First, config has several necessary fields:
- `train`: It contains training related parameters and is a dict type.
- `model`: Model used by the task, specify the generation method in the file, due to the characteristics of lazycall, the model     will be generated at runtime.

> - `optim`: Optimizer related. If not defined, the default will be used.

Reviewer 4: `if not defined xxx` -> `default to AdamW`就行

> - `lr_scheduler`: Related to learning rate, If not defined, the default will be used.

Reviewer 4: 这里同理，改为`Default to warmup cosine decay lr-scheduler`

2. Prepare a model file(such as [model.py](https://github.com/Oneflow-Inc/libai/blob/main/projects/QQP/modeling/model.py)) which contains:
    - Build related models in this file, the construction method is similar to OneFlow.
    - Because Libai will set up a static diagram by default, the calculation of loss needs to be inside the model.
    - The function `forward` must return a dict.
    - When defining a tensor in the model, you need to use `to_global`, turn tensor into a consistent pattern.
    - When defining layers, you can import them directly from `libai.layers`, because it already have SBP defined.

3. Prepare a dataset file(such as [dataset.py](https://github.com/Oneflow-Inc/libai/tree/main/projects/QQP/dataset)) which contains:
    - Build `Dataset` in this file, the construction method is similar to OneFlow.
    - The difference is that we need to use `DistTensorData` and `Instance`.
    - The shape of each batch must be consistent.
    - The return `key`  
    - In `__getitem__` function, the `key` returned by the method must be consistent with the parameter name of the 'forward' function in the 'model'.

## Main Function Entry

> [tools.train_net.py](https://github.com/Oneflow-Inc/libai/blob/main/tools/train_net.py) is the default main function entry provided in LiBai, so we only need to rewrite some functions based on this file.

Reviewer 2: 这个例子重写了哪些函数呢，没看出来

Author: 哦哦，我改一下介绍，之前重写的evaluator是放在这的，后来不需要了

Here is an example:

```python
import numpy as np
from libai.config import LazyConfig, default_argument_parser, try_get_key
from libai.trainer import DefaultTrainer, default_setup
from libai.utils import distributed as dist
from libai.utils.checkpoint import Checkpointer

def main(args):
    cfg = LazyConfig.load(args.config_file)
    cfg = LazyConfig.apply_overrides(cfg, args.opts)
    default_setup(cfg, args)
    trainer = DefaultTrainer(cfg)
    return trainer.train()

if __name__ == "__main__":
    args = default_argument_parser().parse_args()
    main(args)
```

## Build Model

> Preparing a model file is simple, a `model.py` example:

Reviewer 2: 这里感觉写的太简单了，需要加个链接并指引用户去看怎么 import torch 的模型  https://github.com/Oneflow-Inc/libai/pull/198，

Reviewer 2: 或者是给个具体模型的链接

```python
from oneflow import nn

class MyModel(nn.Module):
    def __init__(self, cfg):
        ...

    def forward(self, input_ids, attention_mask):
        ...
        return {'loss': loss}
```

## Build Datasets

> Preparing a dataset file is simple, a `datasets.py` example:

Reviewer 2: 这里一样，给个libai内置数据集的链接，而且每个函数具体的功能感觉也要介绍一下

> All imported modules must take LiBai as the root directory, otherwise, the saved `yaml` file will not be able to save the correct path of the module, resulting in an error when reading `yaml`, so the experiment cannot be reproduced.

After building the `config.py`, if we want to get the corresponding fields in the project, we just need to access like `cfg.my_cfg.***`.

## Start Training
The `train.sh` file contains some parameters, such as `GPUS`, `NODE`, etc.

```bash
#!/usr/bin/env bash
FILE=$1
CONFIG=$2
GPUS=$3
NODE=${NODE:-1}
NODE_RANK=${NODE_RANK:-0}
ADDR=${ADDR:-127.0.0.1}
PORT=${PORT:-12345}

python3 -m oneflow.distributed.launch \
--nproc_per_node $GPUS --nnodes $NODE --node_rank $NODE_RANK --master_addr $ADDR --master_port $PORT \
$FILE --config-file $CONFIG ${@:4}
```

After the above modules are built, we can start training with single gpu.

> Config can support both `py` files and generated `yaml` files.

```bash

> bash projects/my_projects/train.sh train_net.py projects/my_projects/config.py 1

Reviewer: train_net.py 应该是 `tools/train_net.py`

Author: okok
